### PR TITLE
wip: Cargo udeps

### PR DESCRIPTION
**Pull request by ryoqun**
#### Problem

seems `cargo-udeps` got improved a lot since when i first gave a shot to it (around 3 years ago): https://github.com/solana-labs/solana/pull/8137#issuecomment-584836171

#### Summary of Changes

Let's pull it in our ci pipeline. 

this can contribute for crate maintenance burden but it turned out we have few stales debs... well, i expecting such delight as happened with the ages-ago's #8137

Fixes #

### CODE REVIEW

**Comment by CriesofCarrots** [comment]
I suppose CI told you this by now, but those `anyhow` dev declarations are for docs examples.

**Comment by ryoqun** [comment]
cc: @ilya-bobyr (you might be interested in this pr as well)

File: storage-bigtable/Cargo.toml [hunk 1]
```diff
@@ -41,10 +41,6 @@ zstd = { workspace = true }
 # statically...
 [target."cfg(not(windows))".dependencies]
 openssl = { workspace = true, features = ["vendored"] }
-# ...except on Windows to avoid having to deal with getting CI past a build-time
-# Perl dependency
-[target."cfg(windows)".dependencies]
-openssl = { workspace = true, features = [] }
```
Comment (ryoqun) [inline storage-bigtable/Cargo.toml:47]: todo

File: install/Cargo.toml [hunk 2]
```diff
@@ -44,3 +43,6 @@ winreg = { workspace = true }
 
 [package.metadata.docs.rs]
 targets = ["x86_64-unknown-linux-gnu"]
+
+[package.metadata.cargo-udeps.ignore]
+normal = ["serde_yaml_08"]
```
Comment (ilya-bobyr) [inline install/Cargo.toml:48]: Not sure if this is how it should be, but the added line seems to be telling `cargo-udeps` to ignore `serde_yaml_08`, but, at the same time, import of `serde_yaml_08` is removed in the same PR.
Or is this just WIP?

File: sdk/program/Cargo.toml [hunk 3]
```diff
@@ -85,3 +82,6 @@ crate-type = ["cdylib", "rlib"]
 
 [features]
 default = []
+
+[package.metadata.cargo-udeps.ignore]
+development = ["anyhow"]
```
Comment (ilya-bobyr) [inline sdk/program/Cargo.toml:87]: Would it make sense to leave a comment explaining the reasoning for the exception?

**Comment by ilya-bobyr** [comment]
From the maintenance standpoint, `cargo-udeps` seems like a great tool!

File: install/Cargo.toml (hunk 2 above)
Comment (ryoqun) [reply install/Cargo.toml:48]: yep. as this pr title says... ;)

File: install/src/config.rs [hunk 4]
```diff
@@ -62,34 +62,9 @@ impl Config {
...
-            } else {
-                eprintln!("restoration succeeded!");
-            }
-        } else {
-            eprintln!("config upgrade succeeded!");
-        }
-        result
+    fn try_migrate_08(_config_file: &str) -> Result<Self, io::Error> {
+        let p = || panic!();
+        p()
```
[the remark this answers is not in the document]
Comment (ryoqun) [reply install/src/config.rs:67]: this is quick and dirty way to go past our strict ci-checks (clippy there) lol

[the remark this answers is not in the document]
Comment (ryoqun) [reply install/src/config.rs:67]: oh, also, i need to investigate the yaml situation later in depth.

File: sdk/program/Cargo.toml (hunk 3 above)
Comment (ryoqun) [reply sdk/program/Cargo.toml:87]: will do. :)

File: ci/test-checks.sh [hunk 5]
```diff
@@ -29,6 +29,8 @@ EOF
   fi
 )
 
+_ cargo "+${rust_nightly}" install cargo-udeps --locked
```
Comment (ryoqun) [inline ci/test-checks.sh:32]: todo: move this out of this pr like #31332

File: sdk/program/Cargo.toml [hunk 6]
```diff
@@ -85,3 +83,7 @@ crate-type = ["cdylib", "rlib"]
 
 [features]
 default = []
+
+[package.metadata.cargo-udeps.ignore]
+normal = ["tiny-bip39"]
```
Comment (ryoqun) [inline sdk/program/Cargo.toml:88]: todo: needs to debug why removing `"tiny-bip39"` could broke the wasm.....

**Comment by ilya-bobyr** [comment]
Revive?
Or is it too out of date to be useful anymore?

**Comment by ryoqun** [comment]
> Revive? Or is it too out of date to be useful anymore?

hehe, i just revived this by reopening and merging master branch and it detected more https://github.com/solana-labs/solana/pull/31320/commits/12b46f3422c29870b2ebc99a33cc3cea81bfdd82

i think `cargo udeps` are useful. however, I'm seeing obvious false-negatives at #32508 in `solana-test-validator-cli`...

**Pull request closed without merging**